лаба5

Buratishkin opened the pull request.

Buratishkin committed:
добавил лабу

Buratishkin committed:
добавил лабу

bltpzdc commented on commands/AbstractCommand.java:
```diff
@@ -0,0 +1,45 @@
+import interfaces.Elementable;
+
+/** Абстрактный класс для всех команд. */
+public abstract class AbstractCommand implements Command, Elementable, Argumentable {
```
почему `AbstractCommand` имплементирует `Argumentable` и `Elementable`? У тебя не все команды, наследуемые от `AbstractCommand` соответствуют этим интерфейсам.

Buratishkin replied:
Мне кажется я вообще могу отказаться от этих интерфейсов и просто оставить эти методы в `AbstractCommand`. Изначально я делал их, чтобы в `CommandHandler`, когда определяется команда(переменная `currentCommand`), я мог вызвать эти методы для обновления поля коллекции `UpdateTime`  и для понимания нужно ли мне в выполнение команды передавать аргумент или нет.

bltpzdc commented on managers/EnvManager.java:
```diff
@@ -0,0 +1,35 @@
+
+/** Менеджер для взаимодействия с переменной окружения */
+public class EnvManager {
+  private static String env = System.getenv("collection");
```
Зачем хардкодить сразу? По семантике `EnvManager` должен любую переменную окружения уметь разрешать, а не только которую ты указал

bltpzdc commented on managers/FileManager.java:
```diff
@@ -0,0 +1,102 @@
+public class FileManager {
+  private final String inputFileName;
+  private final CityManager cityManager;
+  private final ValidationManager validationManager = new ValidationManager();
```
Dependency inversion из SOLID'а тут был проигнорирован?

bltpzdc commented on managers/FileManager.java:
```diff
@@ -0,0 +1,102 @@
+  /** Сохраняет коллекцию в файл */
+  public void saveCollection() {
+    try {
+      XmlMapper mapper = new XmlMapper();
```
Это можно сделать полем класса и передавать через dependency injection, чтобы не привязываться жестко к реализации и сделать класс универсальным

bltpzdc commented on managers/InputManager.java:
```diff
@@ -0,0 +1,238 @@
+
+/** Менеджер для взаимодействия с пользовательским вводом */
+public class InputManager {
+  private final ValidationManager variableValidator = new ValidationManager();
```
добавь dependency injection

bltpzdc commented on managers/InputManager.java:
```diff
@@ -0,0 +1,238 @@
+import java.util.Scanner;
+
+/** Менеджер для взаимодействия с пользовательским вводом */
+public class InputManager {
```
Вообще как будто бы класс нуждается в переработке полностью. Тут просто привязка жесткая к классу `City`. Либо делай абстракцию нормальной, либо уточняй название, что этот класс только для работы с `City`

bltpzdc commented on managers/ValidationManager.java:
```diff
@@ -0,0 +1,325 @@
+import java.time.LocalDate;
+
+/** Менеджер для ограничения переменных */
+public class ValidationManager {
```
Претензия та же, что и к `InputManager`

bltpzdc commented on service/IdCreator.java:
```diff
@@ -0,0 +1,62 @@
+import managers.CityManager;
+
+/** Класс для генерации id */
+public class IdCreator {
```
Я не совсем понимаю, почему ты решил вынести функцию выделения айди в `IdGenerator`. Можешь мне в треде пояснить в принципе

Buratishkin replied:
этот класс выдает минимальный свободный id, то есть если у меня в коллекции id = [1, 3, 4], то следующему элементу коллекции присвоется id = 2

bltpzdc reviewed (changes requested):
В основном проблема в том, что этот код труднорасширяем. Представь ситуацию, что тебе нужно сделать коллекцию `Foo` вместо `City`. Сколько придется приложить усилий, чтобы это смогло работать? Придется переписывать `InputManager`, `ValidationManager` и `CityManager`. Продумай логику, чтобы переход на новую коллекцию был менее больным

Buratishkin committed:
вроде добавил абстракцию

Buratishkin committed:
добавил интерактивный скрипт мод

Структура скрипта:
    Каждая команда начинается с новой строки. Для получения списка существующих команд напишите help.
    Если команда добавляет элемент в коллекцию, то перед значениями полей должно быть 4 пробела.
    Если значение полей меньше количества, нужных для создания класса, то недостающие поля нужно будет вводить вручную.
    Если больше - лишние поля просто отбрасываются.
Если возникает рекурсия - выполнение скрипта прекращается.

Buratishkin committed:
удалил бесполезные интерфейсы

Buratishkin committed:
подправил баги
